src/fs_io.rs

```diff
@@ -0,0 +1,803 @@
+
+use path::{Path, PathBuf};
+
+/**
+ * These traits and structs interface out our interactions with the fs package.
+ * By using these abstractions, we can tightly control our access to the filesystem,
+ * handle any future changes in the fs API, optionally extend the API, and fully
+ * mock out fs access for testing purposes.
+ *
+ * In general, it's best to match the existing fs api without extension. This way
+ * we can use fs docs knowing that *Container traits are pass through functions to the
+ * underlying fs crate.
+ *
+ * If a pass through function does not exist for the method you need, simply add it to
+ * the trait and add a pass-through implementation.
+ */
+pub trait MetadataContainer {
+    fn is_file(&self) -> bool;
+    fn is_dir(&self) -> bool;
+}
+
+pub struct Metadata(fs::Metadata);
+
+impl MetadataContainer for Metadata {
+    fn is_file(&self) -> bool {
+        self.0.is_file()
+    }
+
+    fn is_dir(&self) -> bool {
+        self.0.is_dir()
+    }
+}
+
+pub trait DirEntryContainer {
+    fn path(&self) -> PathBuf;
+}
+
+pub struct DirEntry(fs::DirEntry);
+
+impl DirEntryContainer for DirEntry {
+    fn path(&self) -> PathBuf {
+        self.0.path()
+    }
+}
+
+pub trait ReadDirContainer: Iterator<Item = io::Result<Box<dyn DirEntryContainer>>> {}
+
+pub struct ReadDir(fs::ReadDir);
+
+impl Iterator for ReadDir {
+    type Item = io::Result<Box<dyn DirEntryContainer>>;
+
+    fn next(&mut self) -> Option<io::Result<Box<dyn DirEntryContainer>>> {
+        match self.0.next() {
+            Some(de) => match de {
+                Ok(de) => Some(io::Result::Ok(Box::new(DirEntry(de)))),
+                Err(e) => Some(io::Result::Err(e)),
+            },
+            None => None,
+        }
+    }
+}
+
+impl ReadDirContainer for ReadDir {}
+
+pub trait FsReadWriter {
+    fn read_dir(&self, path: &str) -> io::Result<Box<dyn ReadDirContainer>>;
+    fn metadata(&self, path: &Path) -> io::Result<Box<dyn MetadataContainer>>;
+}
+
+pub struct Fs {}
+
+impl FsReadWriter for Fs {
+    fn read_dir(&self, path: &str) -> io::Result<Box<dyn ReadDirContainer>> {
+        match fs::read_dir(path) {
+            Ok(rd) => io::Result::Ok(Box::new(ReadDir(rd))),
+            Err(e) => io::Result::Err(e),
+        }
+    }
+
+    fn metadata(&self, path: &Path) -> io::Result<Box<dyn MetadataContainer>> {
+        match fs::metadata(path) {
+            Ok(metadata) => io::Result::Ok(Box::new(Metadata(metadata))),
+            Err(e) => io::Result::Err(e),
+        }
+    }
+}
+
+pub struct FilterResult {
+    pub paths: Vec<String>,
+    pub io_errors: Vec<io::Error>,
+}
+
+pub struct FsFilterer<FS: FsReadWriter> {
+    fs_rw: FS,
+}
+
+impl FsFilterer<Fs> {
+    pub fn new() -> FsFilterer<Fs> {
+        FsFilterer { fs_rw: Fs {} }
+    }
+}
+
+impl<FS: FsReadWriter> FsFilterer<FS> {
+    pub fn new_custom(fs_rw: FS) -> FsFilterer<FS> {
+        FsFilterer { fs_rw }
+    }
+
+    pub fn dir(
```

Comment:
Thoughts on adopting a standard method comment style?  Method description, argument descriptions, and return value description.

Reply:
open to any suggestions. we can add a doc to describe codebase standards.

Reply:
Just remembered, no need to add all of that to the comments. Rust's doc generator is excellent and will extract all the info it needs out of the function signature itself. The only comment we need to remember to add is a prose block comment above the function with all the additional information/explanation you think necessary.